.changeset/wicked-squids-do.md

```diff
@@ -0,0 +1,131 @@
+---
+"@latticexyz/block-logs-stream": patch
+"@latticexyz/cli": patch
+"@latticexyz/common": minor
+"@latticexyz/dev-tools": patch
+"@latticexyz/store-sync": patch
+"@latticexyz/store": major
+"@latticexyz/world": major
+"create-mud": patch
+---
+
+- The external `setRecord` and `deleteRecord` methods of `IStore` no longer accept a `FieldLayout` as input, but load it from storage instead.
+  This is to prevent invalid `FieldLayout` values being passed, which could cause the onchain state to diverge from the indexer state.
+  However, the internal `StoreCore` library still exposes a `setRecord` and `deleteRecord` method that allows a `FieldLayout` to be passed.
+  This is because `StoreCore` can only be used internally, so the `FieldLayout` value can be trusted and we can save the gas for accessing storage.
+
+  ```diff
+  interface IStore {
+    function setRecord(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+      bytes calldata staticData,
+      PackedCounter encodedLengths,
+      bytes calldata dynamicData,
+  -   FieldLayout fieldLayout
+    ) external;
+
+    function deleteRecord(
+      ResourceId tableId,
+      bytes32[] memory keyTuple,
+  -   FieldLayout fieldLayout
+    ) external;
+  }
+  ```
+
+- The `spliceStaticData` method and `Store_SpliceStaticData` event of `IStore` and `StoreCore` no longer include `deleteCount` in their signature.
+  This is because when splicing static data, the data after `start` is always overwritten with `data` instead of being shifted, so `deleteCount` is always the length of the data to be written.
+
+  ```diff
+
+  event Store_SpliceStaticData(
+    ResourceId indexed tableId,
+    bytes32[] keyTuple,
+    uint48 start,
+  - uint40 deleteCount,
+    bytes data
+  );
+
+  interface IStore {
+    function spliceStaticData(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+      uint48 start,
+  -   uint40 deleteCount,
+      bytes calldata data
+    ) external;
+  }
+
+  ```
+
+- All methods that are only valid for dynamic fields (`pushToField`, `popFromField`, `updateInField`, `getFieldSlice`)
+  have been renamed to make this more explicit (`pushToDynamicField`, `popFromDynamicField`, `updateInDynamicField`, `getDynamicFieldSlice`).
```

Comment:
should we make `updateInDynamicField` more like a `spliceDynamicField`?

Reply:
there is a `spliceDynamicField` function too! `updateInDynamicField` was there before but now basically just calls `spliceDynamicField` with the correct `deleteCount` value and does a range check. In fact the range check is not even necessary, since `_spliceDynamicField` does a range check already. Could argue `updateInDynamicField` not really necessary.

Reply:
yeah I am leaning towards us removing, the name was always confusing to me